File: 417_Pacific Atlantic Water Flow.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution{
    public:

    // using dfs 
        int m, n;
    // this denotes cells reachable starting from atlantic and pacific edged cells resp
    vector<vector<bool> > atlantic, pacific;
	vector<vector<int> > res;    
    vector<vector<int> > pacificAtlantic(vector<vector<int>>& heights) {
        if(!size(heights)) return res; // check for null condition
        m = size(heights), n = size(heights[0]); // initializing row and col

    // taking bolean matrices(vector), to see whether a cell that can reach the  atalantic of pacific
        atlantic = pacific = vector<vector<bool> >(m, vector<bool>(n, false));
		// perform dfs from all edge cells (ocean-connected cells)
        for(int i = 0; i < m; i++) dfs(heights, pacific, i, 0), dfs(heights, atlantic, i, n - 1);
        for(int i = 0; i < n; i++) dfs(heights, pacific, 0, i), dfs(heights, atlantic, m - 1, i);             
        return res;
    }
    // define dfs
    void dfs(vector<vector<int> >& mat, vector<vector<bool> >& visited, int i, int j){        
        if(visited[i][j]) return;
        visited[i][j] = true;
		// if cell reachable from both the oceans, insert into final result vector
        if(atlantic[i][j] && pacific[i][j]) res.push_back(vector<int>{i, j});    
		// dfs from current cell only if height of next cell is greater
/*down dir*/  if(i + 1 <  m && mat[i + 1][j] >= mat[i][j]) dfs(mat, visited, i + 1, j); 
/*Up dir*/  if(i - 1 >= 0 && mat[i - 1][j] >= mat[i][j]) dfs(mat, visited, i - 1, j);
/*right dir*/  if(j + 1 <  n && mat[i][j + 1] >= mat[i][j]) dfs(mat, visited, i, j + 1); 
/*left dir*/  if(j - 1 >= 0 && mat[i][j - 1] >= mat[i][j]) dfs(mat, visited, i, j - 1);
    }
};
int main(){

}

File: 417-pacific-atlantic-water-flow/417-pacific-atlantic-water-flow.java
class Solution {
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        int rows = heights.length, cols = heights[0].length;
        boolean[][] pac = new boolean[rows][cols];
        boolean[][] atl = new boolean[rows][cols];
        
        for (int col = 0; col< cols; col++){
            dfs(0, col, rows, cols, pac, heights[0][col], heights);
            dfs(rows-1, col,rows, cols, atl, heights[rows-1][col], heights);
        }
        for (int row = 0; row<rows; row++){
            dfs(row, 0,rows, cols, pac, heights[row][0], heights);
            dfs(row, cols-1,rows, cols, atl, heights[row][cols-1], heights);
        }
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        for (int i = 0; i < rows; i++)
            for (int j = 0; j < cols; j++){
                if (pac[i][j] && atl[i][j])
                    result.add(Arrays.asList(i,j));
            }
        return result;
    }
    private void dfs(int row, int col, int rows, int cols, boolean[][] visited, int prevHeight, int[][] heights){
        if (row < 0 || row >= rows || col < 0 || col >= cols || visited[row][col] || prevHeight > heights[row][col])
            return;
        visited[row][col]= true;
        dfs(row+1, col, rows, cols, visited, heights[row][col], heights);
        dfs(row-1, col, rows, cols, visited, heights[row][col], heights);
        dfs(row, col+1, rows, cols, visited, heights[row][col], heights);
        dfs(row, col-1, rows, cols, visited, heights[row][col], heights);
        
    }
}